### ISSUE TITLE
More Advanced Search metadata consolidation (refactor)

### ISSUE THREAD

**Issue by hepcat72** [Princeton-LSI-ResearchComputing/tracebase#149]
### Feature Inspiration

There are lots of places to edit if you want to add a new search output format.  PR #141 consolidated a lot of it, but there still remain lots of places to edit that would make new output format addition error-prone.

### Feature Description

I would like there to be one place to add a new format.  Currently, that place is inside compositeviews.py (though you do need to add that class to a new class inside forms.py).  The places that need editing to make a single edit location currently involve these files:

- `DataRepo/templates/DataRepo/search/query.html`
- `DataRepo/templates/DataRepo/search/downloads/download.tsv`
- `DataRepo/templates/DataRepo/search/downloads/<format name>.tsv`
- `DataRepo/templates/DataRepo/search/results/display.html`
- `DataRepo/templates/DataRepo/search/results/<format name>.html`

The things which need to be subsumed into compostieviews.py and removed elsewhere are:

- [x] Occurrences of the strings "pgtemplate" and "pdtemplate" in the following files (other than those occurrences linked to specific template files).  *(E.g. Functions that use these in an if/elif should instead use a loop.)*
   - [x] `DataRepo/templates/DataRepo/search/query.html`
   - [x] `DataRepo/templates/DataRepo/search/results/display.html`
- **X** [overly complex*] Occurrences of the strings "pgtemplate" and "pdtemplate" in the following files, by creating generic templates:
   - **X** [overly complex*] `DataRepo/templates/DataRepo/search/downloads/<format name>.tsv`
   - **X** [overly complex*] `DataRepo/templates/DataRepo/search/results/<format name>.html`
   - **X** [overly complex*] `DataRepo/templates/DataRepo/search/downloads/download.tsv`
   - **X** [overly complex*] `DataRepo/templates/DataRepo/search/results/display.html`
- [x] Output format Root models (e.g. the `PeakGroup` and `PeakData` classes)
-  **X** [overly complex*] Field headers in the 2 templates. E.g. Eliminate the `<format name>.*` files by creating 1 file of each type that gets its column headers dynamically
   -  **X** [overly complex*] Add column order to compositeviews
   -  **X** [overly complex*] Create a way to loop through the ordered model fields in a new results html template
   -  **X** [overly complex*] Create a way to loop through the ordered model fields in a new results tsv template
- **X** [overly complex*] Handle field formatting.  *(I could use the strategy I used in branch `compound_view_refactor_generic2` where I created field template files and a default representation when not present.)*
- [x] Results Headers in the templates
- **X** [overly complex*] **OPTIONAL**: The `AdvSearchPeakGroupsForm` and `AdvSearchPeakDataForm` form classes.  I may be able to eliminate this class by creating a base compositeview class that sets up dervied form classes
- [x] Add documentation for adding a new format

\* The hoops you have to jump through to handle variable nested loops in templates ends up making the code really hard to read, so I decided to back off on the 1-stop shop for adding formats.  So I will get it down to eliminating edits to query.html, minimal edits to download.tsv and display.html (just additional elifs), an additional 1-line class in forms.py (and its addition to a loop in AdvSearchForm), and the main files to create/edit: compositeviews.py, and the html and tsv templates.  In fact, I'll add a README for what needs editing.

It would also probably make it easier to do by doing this:

- **X** [create a separate issue] Create a base search class in `views.py` that both the basic and advanced search would be derived from.

Anything overlooked would likely easily be found by adding the third output format.

### Alternatives Considered

none

### Comment

I experimented with adding verbose names as column headers when I tried creating generic templates.  It's possible to extract them from the model objects via custom tags, but I think it could easily be accomplished through context variables.  Field order could also be specified, but custom displays like decimal filters and links would be perhaps over-complicated.  I'll have to think about what is the best balance between readability and one-stop editing.

Note, I also dismissed things like using straight model verbose-names, because "Name" in the context of a detail view is enough, but in a composite view, you need to prepend "Sample" or "Animal" or even fully rename it.

-----

### Assumptions

none

### Requirements

- 1 place to edit (e.g. `compositeviews.py`) to add a new advanced search output format
   - field templates that are used potentially in more than 1 format
   - There will be a single template for each (tsv & html) output format

### Limitations

- Only 1 field template for a table field for inclusion in every composite view, so no custom field views per output format.  E.g. the sample name field will display in 1 way for both the PeakData and PeakGroups formats.  Though note that the variable for the ID used in the links will be different based on the root table, so a separate variable will have to be used using a `with` tag.

### Affected/Changed Components

- `DataRepo/compositeviews.py`
- `DataRepo/views.py`
- `DataRepo/templatetags/customtags.py`
- `DataRepo/templates/search/{results,downloads}/*`

#### GUI Change description

No GUI changes

#### Code Change Description *(Pseudocode optional)*

Basically, I intend to use the same strategy as `compound_view_refactor_generic2`, though the fields will have to include paths and may need to be stored & accessed differently.

This means that each field will have its own template.  If no template exists for it, there will be a default generated view.

#### Tests

No effective changes, so existing tests would work, though they may need to be edited.

**Comment by hepcat72** [comment]
1 thing at a time...  If I want to eliminate coded references to `{pg,pd}template`, then these are the files to edit:
- `DataRepo/templates/DataRepo/search/downloads/download.tsv`
- `DataRepo/templates/DataRepo/search/query.html`
- `DataRepo/templates/DataRepo/search/results/display.html`
...and I would probably have to add code to:
- `DataRepo/compositeviews.py`
to do it.
